### PR TITLE
fetchgit: add record-playback

### PR DESCRIPTION
**Pull request by pmahoney**
(This is just a proof-of-concept; requesting feedback)

Many build scripts assume the ability to run git commands such as
'ls-files' or 'describe'. But unless fetchgit removes `.git`, the source
is non-deterministic (see https://github.com/NixOS/nixpkgs/issues/8567).

This patch arranges for fetchgit to record the output of various
commands at fetch time, and playback that output at build time.

TODO: fetchgit currently requires leaveDotGit when deepClone is used,
the latter giving a functioning `git describe`. But with
record-playback, the `.git` should still be removed.

Example:

```
{ stdenv, fetchgitRecord, record-playback }:

in stdenv.mkDerivation rec {

  src = fetchgitRecord {
    url = "git://someurl";
    rev = "some-revision";
    sha256 = "xxxxxxx";
    recordCommands = [ "ls-files -z" ];
  };

  buildInputs = [ record-playback ];

  ...
}
```

### CODE REVIEW

**Comment by andrewrk** [comment]
Creative solution. I'm impressed.

It might be informative to grep nixpkgs for `leaveDotGit` and start converting packages to use this, and see how it fares.

**Comment by pmahoney** [comment]
Good suggestion @andrewrk, that's a good next step.

Also, @nbp suggested roughly the same record+playback solution in #8567, want to give credit.

File: pkgs/build-support/record-playback/record-command [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  shift
+  PLAYBACK=1
+fi
+
+cmd="$1"; shift
+shortargs="$(printf %s "$*" | tr -d '[:space:]')"
+dir=".record-playback/$cmd/$shortargs"
+
+if [ -n "$PLAYBACK" ]; then
+  printf "playing back \`%s %s' from %s...\n" "$cmd" "$*" "$dir" >&2
```
Comment (nbp) [inline pkgs/build-support/record-playback/record-command]: The command error output might be piped into another command, I will recommend to use a separate fd for than, such as `>&3`, and then we can configure the setup script to forward any output to 3rd file descriptor to the standard error.

File: pkgs/build-support/record-playback/record-command [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#! /bin/sh
+
+if [ "$1" = "--playback" ]; then
+  shift
+  PLAYBACK=1
+fi
+
+cmd="$1"; shift
+shortargs="$(printf %s "$*" | tr -d '[:space:]')"
```
Comment (nbp) [inline pkgs/build-support/record-playback/record-command]: Command lines can potentially be larger than file names, `mkdir -p "$dir"` might fail to record the command line.

**Comment by nbp** [comment]
Nice and generic implementation, I think this code looks good as it is :+1: and that the previous remarks can be addressed in follow-up patches.

**Comment by pmahoney** [comment]
Developing with this is somewhat tedious. Changing the `recordCommands` list does not result in fetchgit re-running, assuming something at the (now out of date) hash is cached. So my workflow is:

* edit sha256 to something bogus
* edit `recordCommands`
* `nix-build`, which fails due to hash mismatch
* copy the actual hash ("instead has xxxx")
* `nix-build` again

Is there a better way?

**Comment by edolstra** [comment]
IMHO this is layering more complexity on top of `fetchgit`, which is just fundamentally broken. For those cases where you absolutely must have a `.git` tree, there is a much simpler solution: just tar the git repository (optionally pruned down to the commits you need), and use fetchurl to download it.

**Comment by pmahoney** [comment]
@edolstra I've got builds that incorporate dozens of (private and public) ruby gems from git repos, and in many cases the gems run `git` commands during build. Some of these change frequently during heavy development cycles. At first glance, it seems more complicated to setup the infrastructure and scripts to create the tarballs, but I'll think about that some more.

**Comment by pmahoney** [comment]
Well this throws a wrench into things, record-playback can't even solve this type of issue: https://github.com/NixOS/nixpkgs/pull/10661#issuecomment-152246516

**Comment by domenkozar** [comment]
`/home/travis/.nox/nixpkgs/lib/customisation.nix:60:31:
anonymous function at /home/travis/.nox/nixpkgs/pkgs/development/interpreters/ruby/bundler-head.nix:1:1 called with unexpected argument ‘ruby’, at `

**Comment by grahamc** [comment]
Is the `record-playback` research worth continuing, or is this project not going to work, due to the issues it doesn't solve?

**Comment by pmahoney** [comment]
This can be closed. It would only solve some issues.  I haven't really had time lately to keep up on this issue, or if other possible solutions have moved forward.

**Comment by marc-hb** [comment]
> For those cases where you absolutely must have a .git tree, there is a much simpler solution: just tar the git repository (optionally pruned down to the commits you need), and use fetchurl to download it.

Great in some cases, very slow and inefficient in other cases.

> It would only solve some issues. 

One size does not fit all. If it did, `leaveDotGit` would not even exist in the first place.

**Pull request closed without merging**